connmgr/banmanager.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package connmgr
+
+import (
+	"net"
+	"sync"
+
+	"github.com/decred/dcrd/peer/v3"
+)
+
+// BanManagerConfig is the configuration struct for the ban manager.
+type BanManagerConfig struct {
+	// DisableBanning represents the status of disabling banning of
+	// misbehaving peers.
+	DisableBanning bool
+
+	// BanThreshold represents the maximum allowed ban score before
+	// misbehaving peers are disconnecting and banned.
+	BanThreshold uint32
+
+	// BanPeer represents a banning callback for cleanup and status
+	// updates purposes on a peer ban.
+	BanPeer func(p *peer.Peer)
+
+	// MaxPeers indicates the maximum number of inbound and outbound
+	// peers allowed.
+	MaxPeers int
+
+	// Whitelist represents the whitelisted IPs of the server.
+	WhiteList []net.IPNet
+}
+
+// banMgrPeer extends a peer to maintain additional state maintained by the
+// ban manager.
+type banMgrPeer struct {
+	*peer.Peer
+
+	isWhitelisted bool
+}
+
+// BanManager represents a peer ban score tracking manager.
+type BanManager struct {
+	cfg       BanManagerConfig
+	peers     map[string]*banMgrPeer
+	banScores map[string]*DynamicBanScore
+	mtx       sync.RWMutex
+}
+
+// NewBanManager initializes a new peer banning manager.
+func NewBanManager(cfg *BanManagerConfig) *BanManager {
+	return &BanManager{
+		cfg:       *cfg,
+		peers:     make(map[string]*banMgrPeer, cfg.MaxPeers),
+		banScores: make(map[string]*DynamicBanScore, cfg.MaxPeers),
+	}
+}
+
+// IsPeerWhitelisted checks if the provided peer is whitelisted.
+func (bm *BanManager) IsPeerWhitelisted(p *peer.Peer) bool {
+	host, _, err := net.SplitHostPort(p.Addr())
+	if err != nil {
+		log.Errorf("Unable to split peer '%s' IP: %v", p.Addr(), err)
+		return false
+	}
+
+	ip := net.ParseIP(host)
+	if ip == nil {
+		log.Errorf("Unable to parse IP '%s'", p.Addr())
+		return false
+	}
+
+	for _, ipnet := range bm.cfg.WhiteList {
+		if ipnet.Contains(ip) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// AddPeer adds the provided peer to the ban manager.
+func (bm *BanManager) AddPeer(p *peer.Peer) {
+	bmp := &banMgrPeer{
+		Peer:          p,
+		isWhitelisted: bm.IsPeerWhitelisted(p),
+	}
+
+	bm.mtx.Lock()
+	bm.peers[p.Addr()] = bmp
+	bm.banScores[p.Addr()] = &DynamicBanScore{}
+	bm.mtx.Unlock()
+}
+
+// LookupPeer returns the associated peer of the provided address and its
+// ban score.
+func (bm *BanManager) LookupPeer(addr string) (*peer.Peer, uint32) {
```

Comment:
Then this could just take the peer instance, be made to return the score, and renamed:

```Go
func (bm *BanManager) BanScore(p *peer.Peer) uint32 {
	bm.mtx.Lock()
	bmp := bm.lookupPeer(p)
	if bmp == nil {
		bm.mtx.Unlock()
	}
	score := bmp.banScore.Int()
	bm.mtx.Unlock()
	return score
}
```